// @mantine/nprogress/cjs/events.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('@mantine/utils');

const [useNavigationProgressEvents, createEvent] = utils.createUseExternalEvents("mantine-nprogress");
const startNavigationProgress = createEvent("start");
const stopNavigationProgress = createEvent("stop");
const resetNavigationProgress = createEvent("reset");
const setNavigationProgress = createEvent("set");
const incrementNavigationProgress = createEvent("increment");
const decrementNavigationProgress = createEvent("decrement");
const completeNavigationProgress = createEvent("complete");

exports.completeNavigationProgress = completeNavigationProgress;
exports.createEvent = createEvent;
exports.decrementNavigationProgress = decrementNavigationProgress;
exports.incrementNavigationProgress = incrementNavigationProgress;
exports.resetNavigationProgress = resetNavigationProgress;
exports.setNavigationProgress = setNavigationProgress;
exports.startNavigationProgress = startNavigationProgress;
exports.stopNavigationProgress = stopNavigationProgress;
exports.useNavigationProgressEvents = useNavigationProgressEvents;
